benchmark/metadata/metadata_test.go

```diff
@@ -0,0 +1,58 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package main
+
+import (
+	"context"
+	"google.golang.org/grpc/metadata"
+	"testing"
+)
+
+func Benchmark_AddingMetadata_WithContextCopying(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		ctx := context.Background()
+		md, _ := metadata.FromOutgoingContext(ctx)
+		send := map[string]string{"k1": "v1", "k2": "v2"}
+		ctx = metadata.NewOutgoingContext(ctx, metadata.Join(metadata.New(send), md))
+	}
+}
+
+func Benchmark_AddingMetadata_AppendToOutgoingContext_WithoutPriorMetadata(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		metadata.AppendToOutgoingContext(context.Background(), "k1", "v1", "k2", "v2")
+	}
+}
+
+func Benchmark_AddingMetadata_AppendToOutgoingContext_PriorMetadata(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		ctx := context.Background()
+		ctx = metadata.NewOutgoingContext(ctx, metadata.MD{})
```

Comment:
This is benchmarking the adding of the initial metadata also, but I don't think that's what you meant to cover?

I would expect this to be:

```go
ctx := metadata.NewOutgoingContext(context.Background(), metadata.MD{})
for n := 0; n < b.N; n++ {
  metadata.AppendToOutgoingContext(ctx, "k1", "v1", ...)
}
```

I also expect this should be no different from the "WithoutPriorMetadata" since AppendToOutgoingContext has to do the same things in both cases.

Reply:
Yes, this can be removed now; I'll do so. In a previous commit, allocation + merging would happen if a non-nil metadata was provided IIRC, which was slower. Thanks for the catch.